{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Outline"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "0. Load Data and Restore results\n",
    "<br />\n",
    "<br />\n",
    "1. Stabilization for linear transformations\n",
    "<br />\n",
    "<br />\n",
    "2. Import packages and Load data\n",
    "<br />\n",
    "<br />\n",
    "3. QC and preprocessing\n",
    "<br />\n",
    "<br />\n",
    "4. Function: identify spatially variable genes\n",
    "<br />\n",
    "<br />\n",
    "5. Function: gene expression enhancement\n",
    "<br />\n",
    "<br />\n",
    "6. Function: characterize functional tissue units"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "SpaGFT is a python package to analyze spatial transcriptomics data via graph Fourier transform. To install SpaGFT, the python version is required to be >= 3.7. You can check your python version by:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'3.8.20'"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import platform\n",
    "platform.python_version()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step 0：Load Data and Restore results\n",
    "Apply SpaGFT on the loaded data and compare SVG list output with the SVG list in the atlas "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import stage\n",
    "from stage_0_setup import stage_0_main as s0\n",
    "# Plotting\n",
    "CELL_DIAMETER = 7 "
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Andersson"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--- Stage 0 Pipeline | Slice: Human_Breast_Andersson_10142021_ST_A1 ---\n",
      "[INFO] Loading CSVs and constructing AnnData...\n",
      "[SUCCESS] AnnData object created: 346 spots x 15045 genes.\n",
      "\n",
      "[INFO] Fetching reference SVG list for SpaGFT...\n",
      "Loading SVG list from local file: C:\\mnt\\data\\SpaGFT-Research\\results\\atlas_results\\Human_Breast_Andersson_10142021_ST_A1_SpaGFT_SVG.csv\n",
      "[SUCCESS] Retrieved 100 reference SVGs.\n",
      "\n",
      "[INFO] Running local SpaGFT implementation...\n",
      "Obatain the Laplacian matrix\n",
      "The precalculated low-frequency FMs are USED\n",
      "The precalculated high-frequency FMs are USED\n",
      "Graph Fourier Transform finished!\n",
      "svg ranking could be found in adata.var['svg_rank']\n",
      "The spatially variable genes judged by gft_score could be found \n",
      "          in adata.var['cutoff_gft_score']\n",
      "Gene signals in frequency domain when detect svgs could be found\n",
      "          in adata.varm['freq_domain_svg']\n",
      "\n",
      "[INFO] Finished running local SpaGFT.\n",
      "\n",
      "========================================\n",
      "RESULTS COMPARISON (Top 100 SVGs)\n",
      "========================================\n",
      "Jaccard Index: 0.7857\n",
      "Overlap: 88\n",
      "========================================\n"
     ]
    }
   ],
   "source": [
    "results_andersson = s0.run_stage_0_pipeline(slice_id=\"Human_Breast_Andersson_10142021_ST_A1\", \n",
    "                         tech=\"ST\")"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Maynard"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--- Stage 0 Pipeline | Slice: Human_Brain_Maynard_02082021_Visium_151676 ---\n",
      "[INFO] Loading CSVs and constructing AnnData...\n",
      "[SUCCESS] AnnData object created: 3460 spots x 33538 genes.\n",
      "\n",
      "[INFO] Fetching reference SVG list for SpaGFT...\n",
      "Attempting to fetch SVG list for Slide: 'Human_Brain_Maynard_02082021_Visium_151676' using Method: 'SpaGFT'...\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "9802414e4a3045c084cf7b82d7595ae5",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "(…)ard_02082021_Visium_151676_organized.csv: 0.00B [00:00, ?B/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Successfully loaded 782 rows.\n",
      "Saved local copy to: C:\\mnt\\data\\SpaGFT-Research\\results\\atlas_results\\Human_Brain_Maynard_02082021_Visium_151676_SpaGFT_SVG.csv\n",
      "[SUCCESS] Retrieved 100 reference SVGs.\n",
      "\n",
      "[INFO] Running local SpaGFT implementation...\n",
      "Obatain the Laplacian matrix\n",
      "The precalculated low-frequency FMs are USED\n",
      "The precalculated high-frequency FMs are USED\n",
      "Graph Fourier Transform finished!\n",
      "svg ranking could be found in adata.var['svg_rank']\n",
      "The spatially variable genes judged by gft_score could be found \n",
      "          in adata.var['cutoff_gft_score']\n",
      "Gene signals in frequency domain when detect svgs could be found\n",
      "          in adata.varm['freq_domain_svg']\n",
      "\n",
      "[INFO] Finished running local SpaGFT.\n",
      "\n",
      "========================================\n",
      "RESULTS COMPARISON (Top 100 SVGs)\n",
      "========================================\n",
      "Jaccard Index: 0.9231\n",
      "Overlap: 96\n",
      "========================================\n"
     ]
    }
   ],
   "source": [
    "results_maynard = s0.run_stage_0_pipeline(slice_id=\"Human_Brain_Maynard_02082021_Visium_151676\",\n",
    "                        tech = \"Visium\")"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Breast 10x"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--- Stage 0 Pipeline | Slice: Human_Breast_10X_06092021_Visium ---\n",
      "[INFO] Loading CSVs and constructing AnnData...\n",
      "[SUCCESS] AnnData object created: 2518 spots x 17943 genes.\n",
      "\n",
      "[INFO] Fetching reference SVG list for SpaGFT...\n",
      "Attempting to fetch SVG list for Slide: 'Human_Breast_10X_06092021_Visium' using Method: 'SpaGFT'...\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "46ab1bc92fc24a4b8e38d4e5c82d5625",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "(…)Breast_10X_06092021_Visium_organized.csv: 0.00B [00:00, ?B/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Successfully loaded 996 rows.\n",
      "Saved local copy to: C:\\mnt\\data\\SpaGFT-Research\\results\\atlas_results\\Human_Breast_10X_06092021_Visium_SpaGFT_SVG.csv\n",
      "[SUCCESS] Retrieved 100 reference SVGs.\n",
      "\n",
      "[INFO] Running local SpaGFT implementation...\n",
      "Obatain the Laplacian matrix\n",
      "The precalculated low-frequency FMs are USED\n",
      "The precalculated high-frequency FMs are USED\n",
      "Graph Fourier Transform finished!\n",
      "svg ranking could be found in adata.var['svg_rank']\n",
      "The spatially variable genes judged by gft_score could be found \n",
      "          in adata.var['cutoff_gft_score']\n",
      "Gene signals in frequency domain when detect svgs could be found\n",
      "          in adata.varm['freq_domain_svg']\n",
      "\n",
      "[INFO] Finished running local SpaGFT.\n",
      "\n",
      "========================================\n",
      "RESULTS COMPARISON (Top 100 SVGs)\n",
      "========================================\n",
      "Jaccard Index: 0.9608\n",
      "Overlap: 98\n",
      "========================================\n"
     ]
    }
   ],
   "source": [
    "results_breast_10x = s0.run_stage_0_pipeline(slice_id=\"Human_Breast_10X_06092021_Visium\",\n",
    "                        tech = \"Visium\")"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### GS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--- Stage 0 Pipeline | Slice: GSE213688_GSM6592060 ---\n",
      "[INFO] Loading CSVs and constructing AnnData...\n",
      "[SUCCESS] AnnData object created: 1295 spots x 36601 genes.\n",
      "\n",
      "[INFO] Fetching reference SVG list for SpaGFT...\n",
      "Loading SVG list from local file: C:\\mnt\\data\\SpaGFT-Research\\results\\atlas_results\\GSE213688_GSM6592060_SpaGFT_SVG.csv\n",
      "[SUCCESS] Retrieved 100 reference SVGs.\n",
      "\n",
      "[INFO] Running local SpaGFT implementation...\n",
      "Obatain the Laplacian matrix\n",
      "The precalculated low-frequency FMs are USED\n",
      "The precalculated high-frequency FMs are USED\n",
      "Graph Fourier Transform finished!\n",
      "svg ranking could be found in adata.var['svg_rank']\n",
      "The spatially variable genes judged by gft_score could be found \n",
      "          in adata.var['cutoff_gft_score']\n",
      "Gene signals in frequency domain when detect svgs could be found\n",
      "          in adata.varm['freq_domain_svg']\n",
      "\n",
      "[INFO] Finished running local SpaGFT.\n",
      "\n",
      "========================================\n",
      "RESULTS COMPARISON (Top 100 SVGs)\n",
      "========================================\n",
      "Jaccard Index: 0.8182\n",
      "Overlap: 90\n",
      "========================================\n"
     ]
    }
   ],
   "source": [
    "results_breast_10x = s0.run_stage_0_pipeline(slice_id=\"GSE213688_GSM6592060\",\n",
    "                        tech = \"Visium\")"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step 1：Stabilization for linear transformations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "my_spagft_env",
   "language": "python",
   "name": "my_spagft_env"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.20"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
