{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Outline"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "0. Load Data and Restore results\n",
    "<br />\n",
    "<br />\n",
    "1. Stabilization for linear transformations\n",
    "<br />\n",
    "<br />\n",
    "2. Import packages and Load data\n",
    "<br />\n",
    "<br />\n",
    "3. QC and preprocessing\n",
    "<br />\n",
    "<br />\n",
    "4. Function: identify spatially variable genes\n",
    "<br />\n",
    "<br />\n",
    "5. Function: gene expression enhancement\n",
    "<br />\n",
    "<br />\n",
    "6. Function: characterize functional tissue units"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'3.8.20'"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import platform\n",
    "platform.python_version()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step 0：Load Data and Restore results\n",
    "Apply SpaGFT on the loaded data and compare SVG list output with the SVG list in the atlas "
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "SpaGFT is a python package to analyze spatial transcriptomics data via graph Fourier transform. To install SpaGFT, the python version is required to be >= 3.7. You can check your python version by:"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 0.1 Update me - global parameters"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Global Parameters - Update all of them before running\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Naming\n",
    "PATH_TO_DATA = r'data/stimage_data_downloads' # data main folder, divided to 3 subfolder: image, gene_exp and coord.\n",
    "PATH_TO_RESULTS = './results/stimage_data/'\n",
    "COORDS_COLUMNS = ['yaxis', 'xaxis'] # name of coordinates columns (e.g. ['x', 'y', 'z'], ['X', 'Y'])\n",
    "CELL_COLUMN = 'cell' # name of cell ids column\n",
    "GENE_COLUMN = 'gene' # name of gene columns\n",
    "TISSUE_NAME = 'Human_Breast_Andersson_10142021_ST_A1' # suit to the tissue name in data file\n",
    "\n",
    "# Scaling\n",
    "CELL_DIAMETER = 7 # For plotting\n",
    "\n",
    "# parameters for functional analysis\n",
    "ORGANISM = 'Human'\n",
    "GENE_SETS = ['BioPlanet_2019','GO_Biological_Process_2021']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "if not os.path.exists(PATH_TO_RESULTS):\n",
    "    os.makedirs(PATH_TO_RESULTS)\n",
    "results_folder = PATH_TO_RESULTS\n",
    "coord_mtx = pd.read_csv(os.path.join(PATH_TO_DATA, 'coord', TISSUE_NAME + '_coord.csv'))\n",
    "count_mtx = pd.read_csv(os.path.join(PATH_TO_DATA, 'gene_exp', TISSUE_NAME + '_count.csv'))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "if 'Unnamed: 0' in coord_mtx.columns:\n",
    "    coord_mtx.index = coord_mtx.iloc[:,0]\n",
    "    coord_mtx.index.name = 'cell_id'\n",
    "    coord_mtx.drop(columns='Unnamed: 0', inplace=True)\n",
    "if 'Unnamed: 0' in count_mtx.columns:\n",
    "    count_mtx.index = count_mtx.iloc[:,0]\n",
    "    count_mtx.index.name = 'cell_id'\n",
    "    count_mtx.drop(columns='Unnamed: 0', inplace=True)\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2. Import packages and Load data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "scanpy==1.9.1 anndata==0.9.2 umap==0.5.7 numpy==1.21.5 scipy==1.10.1 pandas==1.4.2 scikit-learn==1.0.2 statsmodels==0.14.1 python-igraph==0.9.10 louvain==0.7.1 pynndescent==0.5.13\n"
     ]
    }
   ],
   "source": [
    "import SpaGFT as spg\n",
    "import numpy as np\n",
    "import scanpy as sc\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "\n",
    "sc.settings.verbosity = 1\n",
    "sc.logging.print_header()\n",
    "sc.settings.set_figure_params(dpi=80, facecolor='white')"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Two key elements are essential for running SpaGFT: the raw gene counts matrix and spatial coordinates of spots. Indeed, the basic object input to SpaGFT should be an anndata object which contains the above two elements.\n",
    "\n",
    "You can also create your datasets which are ```anndata``` objects. Note that the raw count matrix should be found in adata.X, and the spatial coordinates of all spots should be found in ```adata.obs``` or ```adata.obsm```. You can use diverse functions provided by scanpy to create the anndata object. \n",
    "More approaches to load datasets can be found at [Scanpy](https://scanpy.readthedocs.io/en/stable/api.html#reading)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# create the anndata object\n",
    "adata = sc.AnnData(count_mtx)\n",
    "adata.obs.loc[:, COORDS_COLUMNS] = coord_mtx\n",
    "adata.obsm = {'spatial': coord_mtx[COORDS_COLUMNS].values} \n",
    "adata.var_names_make_unique()\n",
    "adata.raw = adata"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3. QC and preprocessing\n",
    "Before the formal analysis, we perform some basic filtering of genes and normalize Visium counts data with the ```normalize_total``` method from Scanpy followed by log-transform."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "# QC\n",
    "sc.pp.filter_genes(adata, min_cells=10)\n",
    "# Normalization\n",
    "sc.pp.normalize_total(adata, inplace=True)\n",
    "sc.pp.log1p(adata)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 4. Function: identify spatially variable genes"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 4.1 Determine the Fourier modes and identify SVGs"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To begin with, determine the number of Fourier modes for detecting SVGs. SpaGFT provides the dermine_frequency_ratio function based on the kneedle algorithm to determine how many FMs used automatically.\n",
    "Note that *ratio_low * $\\sqrt{n}$* and *ratio_high * $\\sqrt{n}$* are the recommended low-frequency and high-frequency FMs, where *$\\sqrt{n}$* is the number of spots in the original dataset."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Obatain the Laplacian matrix\n"
     ]
    }
   ],
   "source": [
    "# determine the number of low-frequency FMs and high-frequency FMs\n",
    "(ratio_low, ratio_high) = spg.gft.determine_frequency_ratio(adata,\n",
    "                                                            ratio_neighbors=1,\n",
    "                                                            spatial_info=COORDS_COLUMNS)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In the following, SpaGFT will transform gene expression signals from the spatial domain to the frequency domain and analyze the frequency signals to identify SVGs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The precalculated low-frequency FMs are USED\n",
      "The precalculated high-frequency FMs are USED\n",
      "Graph Fourier Transform finished!\n",
      "svg ranking could be found in adata.var['svg_rank']\n",
      "The spatially variable genes judged by gft_score could be found \n",
      "          in adata.var['cutoff_gft_score']\n",
      "Gene signals in frequency domain when detect svgs could be found\n",
      "          in adata.varm['freq_domain_svg']\n",
      "The number of SVGs:  897\n",
      "The top 20 SVGs:\n",
      "['IGKC', 'TFF3', 'MUCL1', 'IGFBP2', 'APOE', 'TPT1', 'SERF2', 'MUC1', 'SCD', 'TAGLN2', 'IGHG3', 'KDELR1', 'KRT17', 'LMNA', 'FNBP1L', 'H2AFJ', 'PERP', 'SCGB2A2', 'COPE', 'ADAM15']\n"
     ]
    }
   ],
   "source": [
    "# calculation\n",
    "gene_df = spg.detect_svg(adata,\n",
    "                         spatial_info=COORDS_COLUMNS,\n",
    "                         ratio_low_freq=ratio_low,\n",
    "                         ratio_high_freq=ratio_high,\n",
    "                         ratio_neighbors=1,\n",
    "                         filter_peaks=True,\n",
    "                         S=6)\n",
    "# S determines the  sensitivity of kneedle algorithm\n",
    "# extract spaitally variable genes\n",
    "svg_list = gene_df[gene_df.cutoff_gft_score][gene_df.fdr < 0.05].index.tolist()\n",
    "print(\"The number of SVGs: \", len(svg_list))\n",
    "# the top 20 SVGs\n",
    "print(\"The top 20 SVGs:\")\n",
    "print(svg_list[:20])\n",
    "\n",
    "# save the gene ranking for all genes\n",
    "gene_df.to_csv(os.path.join(results_folder, TISSUE_NAME + '_gene_rank_gft.csv'))\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step 1：Stabilization for linear transformations"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "spagft_env",
   "language": "python",
   "name": "spagft_env"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.20"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
